crates/telemetry/src/lib.rs

```diff
@@ -16,9 +17,35 @@ pub use propagation::inject_trace_context;
 /// Initializes telemetry for Spin using the [tracing] library.
 ///
 /// Under the hood this involves initializing a [tracing::Subscriber] with multiple [Layer]s. One
-/// [Layer] emits [tracing] events to stderr, and another sends spans to an OTEL collector.
+/// [Layer] emits [tracing] events to stderr, another sends spans to an OTel collector, and another
+/// sends metrics to an OTel collector.
 ///
-/// Configuration is pulled from the environment.
+/// Configuration for the OTel layers is pulled from the environment.
+///
+/// Examples of emitting traces from Spin:
+///
+/// ```no_run
+/// #[instrument(name = "span_name", Err(level = Level::INFO), fields(otel.name = "dynamically set name"))]
+/// fn func_you_want_to_trace() {}
+/// ```
+///
+/// Some notes on tracing:
+///
+/// - If you don't want the span to be collected by default emit it at a trace or debug level.
+/// - Make sure you `.instrument(tracing::Span::current())` any spawned tasks so the span context is propagated.
+/// - Use the otel.name attribute to dynamically set the span name.
+/// - Use the Err argument to have instrument automatically handle errors.
+///
+/// Examples of emitting metrics from Spin:
+///
+/// ```no_run
+/// tracing::trace!(monotonic_counter.metric_foo = 1, attribute = "value");
+/// tracing::trace!(counter.metric_bar = -1, attribute = "value");
+/// tracing::trace!(histogram.another_metric = 1.1, attribute = "value");
```

Comment:
It might be good to add simple wrapper macros for these, e.g.:

```suggestion
/// spin_telemetry::monotonic_counter!(metric_foo = 1, attribute = "value");
/// spin_telemetry::counter!(metric_bar = -1, attribute = "value");
/// spin_telemetry::histogram!(another_metric = 1.1, attribute = "value");
```

- Slightly less magical / surprising
- Gives a place to change our mind about the implementation later.

Reply:
Yup I like that idea.